### ISSUE TITLE
gtk3 User Variables

### ISSUE THREAD

**Issue by B1tl0rd** [asbru-cm/asbru-cm#282]
Hi, there is an error in the variable name field.

![image](https://user-images.githubusercontent.com/12401985/69184636-fe6b5880-0af3-11ea-80e3-dae925402f04.png)
gtk3   a4056b4

**Comment by gfrenoy** [comment]
Sorry, not sure to understand what is wrong ... Can you elaborate on what you would expect ?  Thanks.

**Comment by B1tl0rd** [comment]
hi, shouldn't it be like that? as the global variables?
![image](https://user-images.githubusercontent.com/12401985/69533603-d8f8a780-0f56-11ea-876e-4493cc74307b.png)

**Comment by gfrenoy** [comment]
So it is more a change of the way it works?

I don't use any variable so I'm not even sure how this is used and don't understand why it has been done differently for connection variables and global variables...

Does anyone has an idea ?  And what could be the consequences of this change ?

**Comment by B1tl0rd** [comment]
That's right, sorry, I explained badly. The idea is to be able to identify each one because if you have several identified by numbers in different nodes, you will not remember that they refer only by numbers.

**Comment by hanspr** [comment]
After developing the new interface to KeePassXC, I understand why this logic was applied, the problem is that all the variables : local, global, etc. Have the same search and replace pattern

`<.+?:.+>`

So you can not tell de difference between : `V[1]:0` and `mysql:something`

To solve this, the replacements are done in order of more specific to least specific

```
<V[\d+]:.+>
<G[\d+]:.+>
...
<.+?:.+>
```

In my case moved the KeePassXC patterns to:

`<.+?\|.+>`

## Needed solution

This means, that all this variables will have to be redesigned, and then brake compatibility with any previous of them on all the existing users configurations. To be able to solve this, all patterns would have to be redesigned to something like

`<.+?:G:.+>`
`<.+?:L:.+>`
`<.+?:X:.+>`

So the first part can be any name and not being confused with any other type.

As a conclusion 3 things have to be implemented:

* Change the pattern for each type of variables
* Change the code in all impacted modules : PACEdit, pac_conn, PACUtils, PACExec, PACPrePostExpect,PACExpect
* Create a script to help transform the already created variables

This is a high impact change.

**Comment by merikz** [comment]
I made this feature, named user variables, in pac years ago but I never sent it upstream as there were no receiver of PRs. I have since stopped using variables, because my use case changed.
I'm pretty sure my code isn't compatible with gtk3, since part of it was an OO refactoring of some of the widget handling code, because I hate the copy+paste coding there were so much of in pac. (And probably still are in asbru, unless the gtk3 rework improved things.)

I can maybe look into what can be done to revive it. I haven't looked at gtk3, so I make no promises. 

Technically, my solution was backward compatible by changing the number of the variable to a text string in the config, and at migration of the config set the string to the number value. So a config migration was triggered at startup if the user variables was an array instead of a hash, IIRC.
By the way, there was a rewrite of the config handling recently. Was there any versioning introduced? That was one of the problems I encountered; knowing how to interpret existing config when you make changes/additions.

**Comment by gfrenoy** [comment]
>  because I hate the copy+paste coding there were so much of in pac.

Some refactoring has been done but there's still a lot do in Ásbrú !

> I can maybe look into what can be done to revive it.

That'd be great.  

And if you have not too much time, publishing what you have done "as is" (and even unfinished) might be a great inspiration for others...

> By the way, there was a rewrite of the config handling recently. Was there any versioning introduced?

No.  I'll create an issue to track this.

**Comment by merikz** [comment]
I'm working on this feature [here](https://github.com/merikz/asbru-cm/tree/NamedUserVars).
It is basically what I used 5 years ago in pac (and started merging into asbru in 2017). It turned out not to be much gtk2/gtk3 differences after all. Biggest problem with rebasing was all white space differences and changed coding standard. 

Current status is that it "works", meaning that it doesn't crash and looks OK in the gui.
More testing is needed and the configuration migration probably needs an overhaul, as well as aligning to #353.
 
@gfrenoy What's the planned time frame for v6.1? 
Should I create a draft PR already now, even if I'm not finished?

**Comment by KlaasT** [comment]
@merikz At the moment we are planning to polish 6.1 at bit more and release shouldn't be that far away. A draft PR against the loki branch is very welcome and we could merge it as soon as it is ready.
